### DISCUSSION TITLE
New Controller in Crazyflie

### DISCUSSION THREAD

**Discussion by AntrikshMehta** [bitcraze/discussions#944] [Q&A]
Hey I'm new to crazyflie drones and drones in general. I read a paper about Uniform Passive Fault-Tolerant Control of a
Quadcopter with One, Two, or Three Rotor Failure and wanted to implement it on the crazyflie drone. For that, I'll need to modify the inbuilt controllers. Any help regarding where to begin from would be much appreciated.

**Comment by knmcguire** [comment]
Hi! This will require some quite some embedded changes to the firmware. We can help out with pointing you in the right direction but there are a lot of things that you will need to try to do yourself. Ofcourse if you get any hiccups 

* Documentation on controllers in the firmware: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/
* The latest addition of a controller in the firmware as a pull request. You can use this as a reference: https://github.com/bitcraze/crazyflie-firmware/pull/1156/
* Be mindful that the folder structure and functions for controllers have been changed/refactored after the last pull request: https://github.com/bitcraze/crazyflie-firmware/pull/1247

Also, mind that the crazyflie's motors are not super powerful... just with 4 and a too-quick takeoff will already saturate them, so I highly doubt that the standard motors can handle the motor failures you described here as these type of controllers is usually done with stronger brushless motors. You can consider trying out the [thrust upgrade kit](https://store.bitcraze.io/collections/bundles/products/thrust-upgrade-bundle-for-crazyflie-2-x.):  Don't want to discourage you from trying this but just letting you know about the limitations.
